scripts/zephyr_module.py

```diff
@@ -215,20 +221,58 @@ def main():
     cmake = ""
     sanitycheck = ""
 
+    Module = namedtuple('Module', ['project', 'meta', 'depends'])
+    # dep_modules is a list of all modules that has an unresolved dependency
+    dep_modules = []
+    # start_modules is a list modules with no depends left (no incoming edge)
+    start_modules = []
+    # sorted_modules is a topological sorted list of the modules
+    sorted_modules = []
+
     for project in projects:
         # Avoid including Zephyr base project as module.
         if project == os.environ.get('ZEPHYR_BASE'):
             continue
 
         meta = process_module(project)
         if meta:
-            kconfig += process_kconfig(project, meta)
-            cmake += process_cmake(project, meta)
-            sanitycheck += process_sanitycheck(project, meta)
+            section = meta.get('build', dict())
+            deps = section.get('depends', [])
+            if not deps:
+                start_modules.append(Module(project, meta, []))
+            else:
+                dep_modules.append(Module(project, meta, deps))
         elif project in extra_modules:
             sys.exit(f'{project}, given in ZEPHYR_EXTRA_MODULES, '
                      'is not a valid zephyr module')
 
+    # This will do a topological sort to ensure the modules are ordered
+    # according to dependency settings.
+    while start_modules:
+        node = start_modules.pop(0)
+        sorted_modules.append(node)
+        node_name = PurePath(node.project).name
+        to_remove = []
+        for module in dep_modules:
+            if node_name in module.depends:
+                module.depends.remove(node_name)
+                if not module.depends:
+                    start_modules.append(module)
+                    to_remove.append(module)
+        for module in to_remove:
+            dep_modules.remove(module)
+
+    if dep_modules:
+        # If there are any modules with unresolved dependencies, then the
+        # modules contains unmet or cyclic dependencies. Error out.
+        modules = [ x.project for x in dep_modules ]
```

Comment:
I imagine the linter will complain about the spaces after [ and before ]? If it doesn't, no worries.

Reply:
Fixed now, eagle eye.